modules/process/manager_stacktraces.go

```diff
@@ -0,0 +1,256 @@
+		if flat {
+			processStacks = append(processStacks, pStack)
+		} else if parent, ok := pidMap[process.ParentPID]; ok {
+			parent.Children = append(parent.Children, pStack)
+		} else {
+			processStacks = append(processStacks, pStack)
+		}
+	}
+
+	goroutineCount := int64(0)
+
+	// Now walk through the "Sample" slice in the goroutines stack
+	for _, sample := range stacks.Sample {
+		stack := &Stack{}
+
+		// Add the labels
+		for name, value := range sample.Label {
+			if name == DescriptionPProfLabel || name == PIDPProfLabel || (!flat && name == PPIDPProfLabel) || name == ProcessTypePProfLabel {
+				continue
+			}
+			if len(value) != 1 {
+				// Unexpected...
+				return nil, 0, fmt.Errorf("label: %s in goroutine stack with unexpected number of values: %v", name, value)
+			}
+
+			stack.Labels = append(stack.Labels, &Label{Name: name, Value: value[0]})
+		}
+
+		stack.Count = sample.Value[0]
+		goroutineCount += stack.Count
+
+		// Now get the processStack for this goroutine sample
+		var processStack *Process
+		if pidvalue, ok := sample.Label[PIDPProfLabel]; ok && len(pidvalue) == 1 {
+			pid := IDType(pidvalue[0])
+			processStack, ok = pidMap[pid]
+			if !ok && pid != "" {
+				ppid := IDType("")
+				if value, ok := sample.Label[PPIDPProfLabel]; ok && len(value) == 1 {
+					ppid = IDType(value[0])
+				}
+				description := "(dead process)"
+				if value, ok := sample.Label[DescriptionPProfLabel]; ok && len(value) == 1 {
+					description = value[0] + " " + description
+				}
+				ptype := "code"
+				if value, ok := sample.Label[ProcessTypePProfLabel]; ok && len(value) == 1 {
+					stack.Labels = append(stack.Labels, &Label{Name: ProcessTypePProfLabel, Value: value[0]})
+				}
+				processStack = &Process{
+					PID:         pid,
+					ParentPID:   ppid,
+					Description: description,
+					Type:        ptype,
+				}
+
+				pidMap[processStack.PID] = processStack
+				added := false
+				if processStack.ParentPID != "" && !flat {
+					if parent, ok := pidMap[processStack.ParentPID]; ok {
+						parent.Children = append(parent.Children, processStack)
+						added = true
+					}
+				}
+				if !added {
+					processStacks = append(processStacks, processStack)
+				}
+			}
+		}
+		if processStack == nil {
+			var ok bool
+			processStack, ok = pidMap[""]
+			if !ok {
+				processStack = &Process{
+					Description: "(unassociated)",
+					Type:        "code",
+				}
+				pidMap[processStack.PID] = processStack
+				processStacks = append(processStacks, processStack)
+			}
+		}
+
+		// Now walk through the locations...
+		for _, location := range sample.Location {
+			for _, line := range location.Line {
+				entry := &StackEntry{
+					Function: line.Function.Name,
+					File:     line.Function.Filename,
+					Line:     int(line.Line),
+				}
+				stack.Entry = append(stack.Entry, entry)
+			}
+		}
+		stack.Description = "(unknown)"
+		if len(stack.Entry) > 0 {
+			stack.Description = stack.Entry[len(stack.Entry)-1].Function
```

Comment:
Similar here, why is the last entry the correct one? Comments will go a long way here for these magic code bits.

Reply:
We want to decide what to give a collapsed stacktrace some kind of name. 

Realistically this could be the top (`sample.Location[0]`) or the bottom of the stack. (`sample.Location[len(...)-1]`)

The top of a goroutine stack in a lot of cases will turn out to be `runtime.select` because the goroutine will be waiting on a select pipe for some work, or it will be a call in some downstream library or within the std library.

The bottom of the stack however, will be the first function that creates the goroutine.

If we have to choose one of the two to describe a goroutine - the bottom is most likely to be of help.

Now... We could do other things - e.g. rely on another goroutine label or walk through the stack and choose the first function from code.gitea.io. There are other ideas but actually the bottom of the stack is probably the simplest thing to do.